File: 0049-group-anagrams/0049-group-anagrams.py
"""
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        groups = {}
        for s in strs:
            key = self.get_hash(s)
            groups[key] = groups.get(key, []) + [s]
        return groups.values()
    
    def get_hash(self, s: str) -> str:
        l = list(s)
        l.sort()
        return "".join(l)
"""

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        groups = {}
        for s in strs:
            occ = [0] * 26
            for l in s:
                occ[ord(l)-ord("a")] += 1
            key = tuple(occ)    
            groups[key] = groups.get(key, []) + [s]
        return groups.values()    

File: 0049-group-anagrams/0049-group-anagrams.go
func groupAnagrams(strs []string) [][]string {
    groups := make(map[string][]string, len(strs))
    for _, s := range strs {
        hash := getHash(s)
        groups[hash] = append(groups[hash], s)
    }
    
    result := make([][]string, 0, len(groups))
    for _, g := range groups {
        result = append(result, g)
    }
    return result
}

func getHash(s string) string {
    rs := []rune(s)
    sort.Slice(rs, func(i, j int) bool {
        return rs[i] < rs[j]
    })
    return string(rs)
}